src/Common/src/Microsoft/Win32/SafeHandles/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Win32.SafeHandles
+{
+    using System;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Allows limited reuse of heap buffers to improve memory pressure. This cache does not ensure
+    /// that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal class SafeHeapHandleCache : IDisposable
+    {
+        private ulong minSize;
+        private ulong maxSize;
+
+        // Want to be able to use in low level libraries so we'll keep an array
+        private SafeHeapHandle[] buffers;
+        private uint bufferCount;
+        private object bufferLock = new object();
+
+        /// <param name="minSize">Smallest buffer size to allocate in bytes.</param>
+        /// <param name="maxSize">The largest buffer size to cache in bytes.</param>
+        /// <param name="maxBuffers">The maximum number of buffers to cache.</param>
+        public SafeHeapHandleCache(ulong minSize = 64, ulong maxSize = 1024 * 4, uint maxBuffers = 0)
+        {
+            this.minSize = minSize;
+            this.maxSize = maxSize;
+            maxBuffers = maxBuffers > 0 ? maxBuffers : (uint)Environment.ProcessorCount * 4;
+            this.buffers = new SafeHeapHandle[maxBuffers];
+        }
+
+        /// <summary>
+        /// Get a HeapHandle
+        /// </summary>
+        public SafeHeapHandle Acquire(ulong minSize = 0)
+        {
+            if (minSize < this.minSize) minSize = this.minSize;
+
+            SafeHeapHandle buffer = null;
+            lock (bufferLock)
+            {
+                if (bufferCount > 0) buffer = buffers[--bufferCount];
+            }
+
+            if (buffer == null)
+                buffer = new SafeHeapHandle(minSize);
+            else if (buffer.ByteLength < minSize)
+                buffer.Resize(minSize);
+
+            return buffer;
+        }
+
+        /// <summary>
+        /// Give a HeapHandle back for potential reuse
+        /// </summary>
+        public void Release(SafeHeapHandle buffer)
+        {
+            if (buffer.ByteLength < this.maxSize)
+            {
+                lock (bufferLock)
+                {
+                    if (bufferCount < buffers.Length)
+                        this.buffers[bufferCount++] = buffer;
+                }
+            }
+        }
+
+        public void Dispose()
+        {
+            this.Dispose(disposing: true);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                SafeHeapHandle[] buffers;
+                uint bufferCount;
+                lock (bufferLock)
+                {
+                    buffers = this.buffers;
```

Comment:
This block here is a bit confusing because of the names (local var same as field name), which underscore-prefixed fields would help with.